# GRADTDA5622 - Big Data Computing Foundations 2
## Case Study 1: CRISP_DM Step 2 Data Understanding (with Exploratory Data Analysis)
Replace the example information below with your real information:
- Semester: Spring/Autumn 20xx (FILL IN)
- Instructor: Mr. X (FILL IN)
- Section: 12345 (FILL IN) (if applicable)
- Student Name(s): Able Baker (FILL IN)
- Student Email(s): baker.12345@osu.edu (FILL IN)
- Student ID(s): 123456789  (FILL IN)

Based on CRISP-DM Template Version V12.
***

***
# Section: Overview
- Insert a short description of the scope of this exercise, any supporting information, etc.
- **(I will fill this in for each assignment - Tom Bihari)**
***

### Assignment Overview
In this assignment, you will analyze an ALTERED copy of the “Hotel Booking Demand” dataset.
- This dataset was pulled on 4/8/22 from: https://www.kaggle.com/datasets/jessemostipak/hotel-booking-demand
- The dataset file is named: **hotel_bookings_with_errors_V1.csv**

**The data has been altered slightly for use in course assignments,etc.:**
- A unique ROW attribute has been added.
- Errors have been added, such as: duplicated records, deleted records, deleted attribute values, erroneous attribute values.
**DO NOT PUBLISH THIS DATASET - it contains intentionally wrong data!**

This assignment covers the **Step 2 - Data Understanding** of the six steps of the **CRISP-DM process model** (and very briefly Step 1 - Business Understanding).

The **objectives** of this assignment are:
- By analyzing and thinking critically about the data, you may identify interesting charcteristics that would be useful when deciding **if** and **how** to use the data.
- The results of this step will help set the stage for Step Three (3), in which you may create a final, cleaned-up dataset that can be used for deeper analysis and model-building in the future.

### Problem Statement
Assume that you are the Director of Data Science for Buckeye Resorts, Inc. (BRI), an international hotel chain.  As is the case for all hotel chains, reservation cancellations cause significant impacts to BRI, in profitability, logistics, and other areas.  The cost to BRI of a cancelled reservation is $500 on average.

- BRI wants to improve (decrease) the cancellation rates at its hotels, using more tailored interventions, based on newly available detailed data.  BRI processes 100,000 reservations per year, so an incremental improvement in cancellation rates would have a significant impact.

- One intervention being considered is to offer a special financial incentive to customers who have reservations, but who are “at risk” of cancellation.  BRI has performed a small pilot test, and has found that offering a $100 discount to a customer who is planning to cancel is effective 30% of the time in inducing the customer not to cancel (by locking in a no-cancellation clause in exchange for the discount).

- BRI leadership has asked your team to analyze the new data, and determine if it is suitable for developing analyses and models that would be effective in predicting which future reservations are likely to be at risk of cancellation, so the aforementioned financial incentive could be offered.

- The head of BRI would then like you attend the upcoming BRI Board of Directors meeting.  She has asked you to present your findings to her and to the BOD, to help them decide whether to go forward with the planned tailored intervention approach, and/or to adjust or abandon the approach.  Your goal is to support the BOD in making a decision.

You decide to look at the dataset and **explore (understand)** its contents.  You **do not** need to create (prepare) a clean dataset from it at this time, and you do not yet know the specific algorithm, formula or model you will use to predict cancellations.  **That will come in later steps.**

### Things To Do
Complete the sections for the first two steps of CRISP-DM in this notebook.  Some intial guidance / sugggestions:
- **Business Understanding**: **Briefly** fill in the cells, using your thoughts on the specific item.  This is strictly for practice, for your benefit - it will not be graded rigorously.  You may leave cells empty or say "I don't know.", etc., if the topic is irrelevant.  **Don't spend a lot of time on it.**
- **Data Understanding**: There are many attributes in this dataset. Once you have done some initial analysis, you may choose to concentrate on ones that seem more promising or useful.  However, you must give evidence and explain why the attributes that you choose to give less attention warrant the lack of attention.

It is essential that you **communicate** your goals, thought process, actions, results, and conclusions to the **audience** that will consume this work.  It is **not enough** to show just the code.  It is not appropriate to show long sections of **unexplained printout**, etc.  Be kind to your readers and provide value to them!

**ALWAYS follow this pattern** when doing **each portion** of the work.  This allows us to give feedback and assign scores, and to give partial credit.  Make it easy for the reader to understand your work.
- Say (briefly) **what** you are trying to do, and **why**.
- Do it (code).
- Show or describe the **result** clearly (and briefly as needed), and explain the significant **conclusions or insights** derived from the results.

**HAVE FUN!**

***
# Section: Setup
- Add any needed imports, helper functions, etc., here.
***

In [1]:
import numpy as np
import pandas as pd

#!pip install matplotlib
import matplotlib
import matplotlib.pyplot as plt
%matplotlib inline
#matplotlib.use('Qt5Agg')

import seaborn as sns

pd.set_option('display.max_columns', 50) #include to avoid ... in middle of display

***
# Section: 1 - Business Understanding
- This step is essential (and occasionally ignored…).  It focuses on understanding the business needs and objectives, and defining a data analysis problem that is relevant and useful to the business, so the right solution is built.

- Consider the following questions and discuss the possible answers to them in the spaces below.
  - It is OK to say that you do not know.  How might you get answers to the questions?
  - You may consider some of the questions to be "not applicable".  If so, why?
  - You may choose to use markdown instead of raw text in the answer cells, to make them "pretty" and easy to read.
***

**Context**
- What is the overview, history, big picture?

**Business Objectives**
- What is the problem or opportunity to be addressed / solved?
- What is its business value?

**Stakeholders**
- Who are they?
- What are their needs?

**General Goals**
- Understand the past/current state (descriptive)?
- Predict the future (predictive)?
- Mine for unknown patterns / insights?
- Answer questions?  What are they?
- Support decisions?  What are they?

**Scope**
- What is the specific scope of the project?
- What, specifically, is NOT in scope?  (What are you not doing that someone might think you are doing?)
Is the scope flexible?  In what way?

**Deliverables**
- Report?
- Reusable code?
- Presentation?

**Success Criteria**
- What, specifically, must be accomplished for this project to be a success?

**Solution Approach**
- Is this a “one time” (throw away) solution?
- Or will the solution become part of repeated process that is used and enhanced over time?

**Quality**
- What level of Correctness or Confidence is needed?
- Is this life-critical or are we looking for ‘directionally correct’ results?

**Constraints**
- Are there legal / regulatory implications (privacy, security, liability, HIPAA, FERPA)?
- Are there architectural constraints?

**Time / Cost**
- What is the deadline or budget?

**Dependencies**
- Is this project dependent on any other project or activity?
- Is any other project or activity dependent on this project?

**Resources Needed**
- Do you potentially need access to specific people, data, etc?

**Other Considerations**
- What other considerations, not covered above, might be relevant?

***
# Section: 2 - Data Understanding
- Perform exploratory data analysis of the dataset by looking at individual attributes and/or combinations of attributes. You should focus on identifying and describing interesting observations and insights that you might uncover from the data.
- You should not simply provide the basic EDA information for all attributes in the data. Instead, you should focus on those that are more interesting or important, and provide some discussion of what you observe. Pay particular attention to potentially interesting bivariate (two-variable) relationships, as well as the relationship between each attribute and the class attribute (if this is a classification problem).
- This step includes:
  - Understanding what data might be available and relevant.
  - Acquiring the data.
  - Understanding the characteristics of the data (e.g., quality, completeness, currency, consistency,…).
  - Understanding how other datasets might enhance or supplement the data (and acquiring that data as well).
  - Possibly performing some initial analysis to build confidence that the overall direction is likely to be helpful to the business problem to be solved.

**NOTE:** Frequently, a first pass is made through the data to fix or remove "obvious" data quality issues (missing values, duplicate records, outliers (some), etc.).  The deeper Data Understanding analysis then is done on the "cleaned" data.  This mitigates the issue of bad data skewing the data analysis results.  **For this example, we will not do this separate first pass.**
***

***
## Section: 2.1 - Describe the meaning and type of data for each attribute.

Identify the type of data for each attribute.
  - Categorical: Nominal, Ordinal
  - Numerical: Interval, Ratio
  - Other: Complex or unstructured data (e.g., text strings, images,...)
  
NOTES:
  - This determines what operations are "meaningful" for each attribute.  For example, you can't add two Ordinal attributes.
  - You can't always tell the attribute types from the data itself.  For example, an Integer attribute could be Ordinal or Interval.

Identify interesting characteristics of the data (e.g., quality, completeness, currency, consistency,…).
  - Is this dataset relevant and appropriate for the goals of this project?
    - If not, what could you do? (find other data to replace or supplement this dataset, etc.)
***

### **Examine the meta-data (e.g., from data dictionary) available for the dataset.**  
#### As a reminder, the meta-data can be found [here](https://www.kaggle.com/datasets/jessemostipak/hotel-booking-demand).

**Discussion:**
- Please fill in a response...

### **Load the Dataset.**

In [ ]:
#OPTIONAL: Load it only once per session, and save it, so we don't waste time.

#NOTE: You may need to change the path to the data file, depending on your setup.

try:
    _ = data_from_source_file_df
    print("Reusing source data")
except:
    print("Loading source data")
    # data_from_source_file_df = pd.read_csv("../shared_Sp23/hotel_bookings_with_errors_V1.csv")
    data_from_source_file_df = pd.read_csv("https://bcoy48.github.io/osu-gradtda-5622-sp25/course_materials/case_studies/1/hotel_bookings_with_errors_V1.csv")
data_df = data_from_source_file_df
print("Done!")

In [3]:
# Print the schema, based on the data in the file.
data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119391 entries, 0 to 119390
Data columns (total 33 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   ROW                             119391 non-null  int64  
 1   hotel                           119391 non-null  object 
 2   is_canceled                     119390 non-null  float64
 3   lead_time                       119391 non-null  int64  
 4   arrival_date_year               119391 non-null  int64  
 5   arrival_date_month              119391 non-null  object 
 6   arrival_date_week_number        119391 non-null  int64  
 7   arrival_date_day_of_month       119391 non-null  int64  
 8   stays_in_weekend_nights         119391 non-null  int64  
 9   stays_in_week_nights            119391 non-null  int64  
 10  adults                          119391 non-null  int64  
 11  children                        119387 non-null  float64
 12  babies          

**Discussion:**
- Please fill in a response...

In [ ]:
# # Create some simple lists for future use
# all_attribute_names = list(data_df)
# print(all_attribute_names)
# class_attribute_name = []
# nominal_attribute_names = []
# ordinal_attribute_names = []
# categorical_attribute_names = nominal_attribute_names + ordinal_attribute_names
# interval_attribute_names = []
# ratio_attribute_names = []
# numerical_attribute_names = interval_attribute_names + ratio_attribute_names

**Discussion:**
- Please fill in a response...

***
## Section: 2.2 - Provide basic statistics for the attributes.
- For example: counts, percentiles, mean, median, standard deviation. The statistics should be relevant for the type of attribute.
***

In [ ]:
# # This is an optional step.  Pandas Profiling is a library that provides a sophisticated profiling report
# #  that can be saved in the form of an interactive HTML file.  See: https://pypi.org/project/pandas-profiling/
# #  It may run very slowly on large datasets.  It is best to limit the number of records to profile, based on
# #  experimentation with your computer.  You can do so by profiling small samples from large datasets.
# #  You also can specify the "minimal=True" argument in newer versions of the library.
# # To install, uncomment and run the following line (one time).
# #!pip install pandas_profiling

# import pandas_profiling
# if data_df.shape[0] > 500:
#     profile = pandas_profiling.ProfileReport(data_df.sample(n=500))  # Limit to a small set of records.
#     #profile = pandas_profiling.ProfileReport(data_df, minimal=True)  # Limit to the simpler profiling.
# else:
#     profile = pandas_profiling.ProfileReport(data_df)
# profile.to_file("./profilereport.html")

# # Set plotting back to interactive mode (pandas_profiling sets it to non-interactive)
# %matplotlib inline

**Discussion:**
- Please fill in a response...

In [4]:
data_df.describe() #calculates statistics for all numerical columns

ROW    is_canceled      lead_time  arrival_date_year  \
count  119391.000000  119390.000000  119391.000000      119391.000000   
mean    59694.000050       0.370408     104.010662        2016.156544   
std     34465.357245       0.482916     106.862967           0.707481   
min         0.000000       0.000000       0.000000        2015.000000   
25%     29846.500000       0.000000      18.000000        2016.000000   
50%     59694.000000       0.000000      69.000000        2016.000000   
75%     89541.500000       1.000000     160.000000        2017.000000   
max    119389.000000       1.000000     737.000000        2017.000000   

       arrival_date_week_number  arrival_date_day_of_month  \
count             119391.000000              119391.000000   
mean                  27.165172                  15.798117   
std                   13.605081                   8.780897   
min                    1.000000                   1.000000   
25%                   16.000000                   8.000000   
50%                   28.000000                  16.000000   
75%                   38.000000                  23.000000   
max                   53.000000                  31.000000   

       stays_in_weekend_nights  stays_in_week_nights         adults  \
count            119391.000000         119391.000000  119391.000000   
mean                  0.927591              2.500297       1.856572   
std                   0.998613              1.908278       0.582185   
min                   0.000000              0.000000       0.000000   
25%                   0.000000              1.000000       2.000000   
50%                   1.000000              2.000000       2.000000   
75%                   2.000000              3.000000       2.000000   
max                  19.000000             50.000000      55.000000   

            children         babies  is_repeated_guest  \
count  119387.000000  119391.000000      119391.000000   
mean        0.103889       0.007949           0.031912   
std         0.398560       0.097436           0.175766   
min         0.000000       0.000000           0.000000   
25%         0.000000       0.000000           0.000000   
50%         0.000000       0.000000           0.000000   
75%         0.000000       0.000000           0.000000   
max        10.000000      10.000000           1.000000   

       previous_cancellations  previous_bookings_not_canceled  \
count           119391.000000                   119391.000000   
mean                 0.087117                        0.137096   
std                  0.844333                        1.497431   
min                  0.000000                        0.000000   
25%                  0.000000                        0.000000   
50%                  0.000000                        0.000000   
75%                  0.000000                        0.000000   
max                 26.000000                       72.000000   

       booking_changes         agent      company  days_in_waiting_list  \
count    119391.000000  103051.00000  6797.000000         119391.000000   
mean          0.221122      86.69487   189.266735              2.321130   
std           0.652303     110.77504   131.655015             17.594648   
min           0.000000       1.00000     6.000000              0.000000   
25%           0.000000       9.00000    62.000000              0.000000   
50%           0.000000      14.00000   179.000000              0.000000   
75%           0.000000     229.00000   270.000000              0.000000   
max          21.000000     535.00000   543.000000            391.000000   

                 adr  required_car_parking_spaces  total_of_special_requests  
count  119391.000000                119391.000000              119391.000000  
mean      101.831089                     0.062517                   0.571366  
std        50.535580                     0.245290                   0.792796  
min        -6.380000                     0.000000          

**Discussion:**
Note that there are some suspicious values above.  These may be data quality issues.  Suggestions:
- Please fill in a response...

***
## Section: 2.3 - Visualize / analyze the most important or interesting attributes using appropriate techniques.
- For each visualization, provide an interpretation explaining why it is appropriate or interesting. What does each visualization tell us?
***

**Discussion:**
Note that there are some suspicious values above.  These may be data quality issues.  Suggestions:
- Please fill in a response...

***
## Section: 2.4 - Verify data quality: explain any missing values, duplicate data, or outliers.
- What, if anything, do you need to do about these? Be specific.
***

### **Missing Attribute Values**

**Discussion:**
Suggested actions:
- Please fill in a response...

### **Duplicate Records**

**Discussion:**
Suggested actions:
- Please fill in a response...

### **Outliers**
- There are many ways to discover and handle outliers, and many resources online.  Here is a nice discussion:
- https://stackoverflow.com/questions/23199796/detect-and-exclude-outliers-in-a-pandas-dataframe

**Discussion:**
Suggested actions:
- Please fill in a response...

***
## Section: 2.5 - Explore the relationships among the attributes.
- You may include or exclude the class attribute as appropriate.
- Use scatter plots, correlation matrices, cross-tabulations, group-wise averages, or other appropriate techniques. Explain and interpret any interesting relationships.
- There are three general cases for 2-attribute relationships:
  - Categorical vs Categorical
  - Categorical vs Numerical
  - Numerical vs Numerical
***

### **Categorical Attribute vs Categorical Attribute**

**Discussion:**
- Please fill in a response...

### **Categorical Attribute vs Numerical Attribute**

**Discussion:**
- Please fill in a response...

### **Numerical Attribute vs Numerical Attribute**

**Discussion:**
- Please fill in a response...

***
## Section: 2.6 - Identify and explain any interesting relationships between the class attribute and the other attributes.
- You may refer to earlier analyses / visualizations (if you included the class attribute above) or create new ones.
***

**Discussion:**
- Please fill in a response...

***
# Section: 3 - Data Preparation
In this step, the “final” target dataset is constructed (although nothing is ever final…), based on the insights gleaned in the Data Understanding step.
This may include:
- Normalizing / standardizing the data.
- Handling missing and erroneous data.
- Transforming or eliminating attributes.
- “Blending” data from multiple sources (with attention to consistency of the various sources).
- Deploying the data to storage in suitable formats, data models / structures.
***

***
## Section: 3.1 - Clean the Data
- Implement any data cleaning steps previously identified.  Please justify.
- Show the effects of that cleaning through the use of appropriate statistics and/or visualizations.
***

**Discussion:**
- Please fill in a response...

***
## Section: 3.2 - Select Attributes
- Optionally, remove attributes that are unneeded, redundant, etc., for the expected uses.
- Which attributes did you decide to remove, if any? Please justify.
***

**Discussion:**
- Please fill in a response...

***
## Section: 3.3 - Create Attributes
- Optionally, create new attributes based on the values of other attibutes.
- Which new attributes did you decide to create, if any? Please justify.
***

**Discussion:**
- Please fill in a response...

***
## Section: 3.4 - Transform Attributes
- Optionally, transform existing attributes (e.g., standardize, normalize, change units of measure, encode) to make them more appropriate for the expected uses.
- Which attributes did you decide to transform, if any? Please justify.
***

**Discussion:**
- Please fill in a response...

***
## Section: 3.5 - Select Records
- Optionally, select a subset of the records, via sampling, etc., to decrease the size of the dataset. Please justify.
***

**Discussion:**
- Please fill in a response...

***
## Section: 3.6 - Other Data Preparation
- Optionally, perform other preparation steps (e.g., to optimize storage, retrieval, etc.). Please justify.
***

**Discussion:**
- Please fill in a response...

***
# Section: 4 - Modeling
In this step, the dataset is analyzed and used to create models and/or answer questions relevant to the business. This may include:
- Developing statistical models for describing or predicting business situations.
- Training data mining (e.g., machine learning) algorithms for classification, clustering, association analysis, outlier detection, and other uses.

NOTE: This step is tightly coupled and iterative with the Data Preparation and Evaluation steps.
***

**Discussion:**
- Please fill in a response...

***
# Section: 5 - Evaluation
In this step, the analyses and/or models are evaluated for quality and relevance / usefulness.

NOTE: This step is tightly coupled and iterative with the Modeling step.
***

**Discussion:**
- Please fill in a response...

***
# Section: 6 - Deployment
In this step, data and resulting models and/or analyses are put to use. This may include:
- Deploying data mining / machine learning algorithms to production.
- Communicating the results of the analyses to business stakeholders, so they can take action.
***

**Discussion:**
- Please fill in a response...

***
# Section: Conclusions
- What did you learn from this exercise?
***

Please fill in a response...